parser-typechecker/src/Unison/Util/Pretty.hs

```diff
@@ -669,25 +710,49 @@ instance Monoid (Pretty s) where
       (_,_) -> pure p1 <> pure p2
 
 data Delta =
-  Delta { line :: !Int, col :: !Int, maxCol :: !Int }
-  deriving (Eq,Ord,Show)
+    -- | The number of columns.
+    SingleLine !Width
+    -- | The number of columns in the first, last, and longest lines.
+  | MultiLine !Width !Width !Width
+  deriving stock (Eq, Ord, Show)
+
+instance Semigroup Delta where
+  SingleLine c <> SingleLine c2 = SingleLine (c + c2)
+  SingleLine c <> MultiLine fc lc mc =
+    let fc' = c + fc
+    in MultiLine fc' lc (max fc' mc)
+  MultiLine fc lc mc <> SingleLine c =
+    let lc' = lc + c
+    in MultiLine fc lc' (max lc' mc)
+  MultiLine fc lc mc <> MultiLine fc2 lc2 mc2 =
+    MultiLine fc lc2 (max mc (max mc2 (lc + fc2)))
 
-instance Semigroup Delta where (<>) = mappend
 instance Monoid Delta where
-  mempty = Delta 0 0 0
-  mappend (Delta l c mc) (Delta 0 c2 mc2) =
-    Delta l (c + c2) (mc `max` mc2 `max` (c + c2))
-  mappend (Delta l _ mc) (Delta l2 c2 mc2) = Delta (l + l2) c2 (mc `max` mc2)
+  mempty = SingleLine 0
+  mappend = (<>)
+
+maxCol :: Delta -> Width
+maxCol = \case
+  SingleLine c -> c
+  MultiLine _ _ c -> c
+
+lastCol :: Delta -> Width
+lastCol = \case
+  SingleLine c -> c
+  MultiLine _ c _ -> c
 
 chDelta :: Char -> Delta
-chDelta '\n' = Delta 1 0 0
-chDelta _ = Delta 0 1 1
+chDelta '\n' = MultiLine 0 0 0
+chDelta _ = SingleLine 1
 
 preferredWidth :: Pretty s -> Width
-preferredWidth p = col (delta p)
+preferredWidth p = lastCol (delta p)
```

Comment:
This might be a bug. I decided not to change the behavior, but it seems weird to say the "preferred width" of something is however wide its last column is.

Still, I'd like to at least have an idea of what changing this to `maxCol` or something might fix, so I left it as-is for now.

Reply:
@pchiusano